### DISCUSSION TITLE
Model fails to detect 1 out of the 5 small object classes

### DISCUSSION THREAD

**Discussion by AlexanderZhou39** [ultralytics/yolov5#10305] [Q&A]
Hello,
I'm rather new to ml in general so I apologize in advance if I ask some dumb questions. 

## The Task
So I'm currently trying to train a model to detect icons in a game screenshot. In particular, there are 5 variations of an icon that are particularly small in the screenshots. 
For example, here is an example screenshot
![fjdjsfjdj](https://user-images.githubusercontent.com/66284119/204108766-2a583564-32dc-4928-8d5f-1bceb8d1844c.png)
And here are 4 of the 5 variations of what I'm trying to detect as example (p3, p4, p5, p6 respectively)
![Screenshot 2022-11-26 at 1 00 37 PM](https://user-images.githubusercontent.com/66284119/204108788-9a83c279-e023-4b29-be55-d6f64029718b.png)
![Screenshot 2022-11-26 at 1 00 43 PM](https://user-images.githubusercontent.com/66284119/204108789-226cc388-abe9-4d8c-acce-fde35ddb699a.png)
![Screenshot 2022-11-26 at 1 33 14 PM](https://user-images.githubusercontent.com/66284119/204109768-92321bb0-a292-4600-bfde-1102607c2359.png)
![Screenshot 2022-11-26 at 1 00 49 PM](https://user-images.githubusercontent.com/66284119/204108790-ac8602ae-4e61-4b14-82dd-65d45f2a56f4.png)

## The Data
Currently, I've labeled each variation as a separate class (p2, p3, p4, p5, p6). I have a tiny data set of screenshots (76 training, 16 val, 16 test). The most common class is p6 with likely 100-200 or more training instances. The least common class is p5 with around 60 training instances. 

## Training Parameters
All images are fit to a 1600 x 800 black background. I use a modified hyp.scratch-low hyperparameter setting where color augmentation is set to a bare minimum (I don't expect much color variation from screenshots), flips are turned off, mosaic is turned off, scale is reduced to 0.2. 
```
lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.01  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.5  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 1.0  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
# anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
hsv_h: 0.003  # image HSV-Hue augmentation (fraction)
hsv_s: 0.1  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.0  # image HSV-Value augmentation (fraction)
degrees: 0.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.2  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # image flip up-down (probability)
fliplr: 0.0  # image flip left-right (probability)
mosaic: 0.0  # image mosaic (probability)
mixup: 0.0  # image mixup (probability)
copy_paste: 0.0  # segment copy-paste (probability)
```
With this and the pre-trained yolov5s.py weights, I was able to get the following results. Batch size was 13. 
![Screenshot 2022-11-23 at 11 28 40 PM](https://user-images.githubusercontent.com/66284119/204109034-c3c21a38-2d4e-4e6e-8c75-1f69e5a62d30.png)
As you can see, it performs excellently on all variations of the icon except for p5. 

## What I've Tried
I've tried training on larger images (up to 2048, 1024) and using different yolov5 models (s6, m, n, n6). I added another 10-30 p5 examples to the training set and did not see better performance (instead the model seemed to start confusing p5 for p4 or p6, and recall remained 0). I tried removing the color and scale augmentations completely. I've also tried increasing batch size to 33 (roughly 1/3 of the data set).

## My Thoughts
At this point, I've thought about merging all the classes together (so under one p label for all) and training a separate classifier or doing template matching on the cropped labels for the sole purpose of distinguishing between the 5. 
I could potentially search for more samples with p5 in them, but I'm not sure of how many I would need to supply, and the other labels already excel despite having relatively few instances. p4 for example has 90 labels and does very well. p5 has 61 and fails completely. It's more like exact template matching than classifying amorphous blobs as cats, so I don't really get why it struggles here. Perhaps p5 is just very similar to p6 and p4? Though it never misclassifies p4 or p6 as p5. 

## General Help
The overarching goal is to be able to detect all of them accurately and quickly. As fast and best as possible, so any solutions that significantly increase the computation cost would not be preferred. 
Like I said, I'm rather clueless as someone who's only recently got into object detection, so it would be greatly appreciated if anybody could offer some insights into the most reasonable course of action. Whether that is to separate the task, add more samples, or something else. 
Thank you for your time 🙏

**Comment by glenn-jocher** [comment]
@AlexanderZhou39 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results.  All your classes already achieve high mAP.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!
